Src/DependencyCollector/Shared/Implementation/ProfilerSqlProcessingBase.cs

```diff
@@ -253,6 +251,80 @@ private object OnBegin(object thisObj)
         }
 
         /// <summary>
+        ///  Common helper for all EndAsync Callbacks.
+        /// </summary>
+        /// <param name="taskObj">Returned task by the async method.</param>
+        /// <param name="thisObj">This object.</param>
+        private void OnEndAsync(object taskObj, object thisObj)
+        {
+            try
+            {
+                Task task = taskObj as Task;
+                if (task == null)
+                {
+                    DependencyCollectorEventSource.Log.NotExpectedCallback(0, "OnEndAsyncSql", "task == null");
+                    return;
+                }
+
+                DependencyCollectorEventSource.Log.EndAsyncCallbackCalled(thisObj.GetHashCode().ToString(CultureInfo.InvariantCulture));
+
+                task.ContinueWith(t =>
+                {
+                    Exception exceptionObj = null;
+                    if (t.IsFaulted)
+                    {
+                        exceptionObj = t.Exception != null && t.Exception.InnerException != null ? t.Exception.InnerException : t.Exception;
+                    }
+
+                    this.OnEnd(exceptionObj, thisObj);
+                });
+            }            
+            catch (Exception ex)
+            {
+                DependencyCollectorEventSource.Log.CallbackError(thisObj == null ? 0 : thisObj.GetHashCode(), "OnEndAsyncSql", ex);
+            }
+        }
+
+        /// <summary>
+        ///  Common helper for all EndAsync Callbacks that should send data only in the case of exception happened.
+        /// </summary>
+        /// <param name="taskObj">Returned task by the async method.</param>
+        /// <param name="thisObj">This object.</param>
+        private void OnEndExceptionAsync(object taskObj, object thisObj)
+        {
+            try
+            {
+                Task task = taskObj as Task;
+                if (task == null)
+                {
+                    DependencyCollectorEventSource.Log.NotExpectedCallback(0, "OnEndExceptionAsync", "task == null");
+                    return;
+                }
+
+                DependencyCollectorEventSource.Log.EndAsyncExceptionCallbackCalled(thisObj.GetHashCode().ToString(CultureInfo.InvariantCulture));
+
+                task.ContinueWith(t =>
+                {
+                    Exception exceptionObj = null;
+                    if (t.IsFaulted)
+                    {
+                        // track item only in case of failure
+                        exceptionObj = t.Exception != null && t.Exception.InnerException != null ? t.Exception.InnerException : t.Exception;
```

Comment:
if IsFaulted is true, then you don't need to check if t.Exception is not null - it will be non-null always as per documentation:
https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.isfaulted(v=vs.110).aspx